src/globus_cli/commands/gcs/endpoint/update.py

```diff
@@ -0,0 +1,244 @@
+from __future__ import annotations
+
+import functools
+import typing as t
+import uuid
+
+import click
+import globus_sdk
+
+from globus_cli.commands.gcs.endpoint._common import GCS_ENDPOINT_FIELDS
+from globus_cli.constants import EXPLICIT_NULL, ExplicitNullType
+from globus_cli.login_manager import LoginManager
+from globus_cli.parsing import CommaDelimitedList, command, endpoint_id_arg
+from globus_cli.parsing.param_types.nullable import IntOrNull
+from globus_cli.termio import TextMode, display
+
+
+class SubscriptionIdType(click.ParamType):
+    def get_type_annotation(self, _: click.Parameter) -> type:
+        return t.cast(type, str | ExplicitNullType)
```

Comment:
The cast here, together with a generic type, surprises me but may reflect a misunderstanding on my part.

Is it possible to avoid the cast by writing the annotation as `-> type[str | ExplicitNullType]` and then returning that? Or `t.Type[str | ExplicitNullType]`?

Reply:
This may be copied from my prior art elsewhere. I can offer some relevant info: `types.UnionType` is not a `type`.

I think this derives in some way from the difference between a `UnionType` at runtime and any "normal" `type`. For example, you can't instantiate a `UnionType`, as in `(int | str)("1")`.

It may be that `type` is the wrong return-type for this, and `click-type-test` might need some tinkering to make it all work nicely. But once the typing folks start talking about "special forms" I generally get scared and want to go back to thinking about the language I know (the runtime one). :wink:

Reply:
> the difference between a `UnionType` at runtime and any "normal" `type`.

_That's_ exactly the kind of misunderstanding that I was expecting. Thanks for clarifying!

Reply:
Looks like no.

I think this has to do with mypy's evaluation of Unions in particular but I'll be the first to admit that I'm simply [following the traditions of my ancestors](https://github.com/globus/globus-cli/blob/main/src/globus_cli/parsing/param_types/json_strorfile.py#L62).

---

**Code**
```python
    def get_type_annotation(self, _: click.Parameter) -> type[str | ExplicitNullType]:
        return str | ExplicitNullType
```

**Mypy Error**
```
error: Incompatible return value type (got "UnionType", expected "type[str] | type[ExplicitNullType]")  [return-value]
```

Reply:
I think `str | ExplicitNullType` indicates an instance of either type will be returned. Perhaps:

```python
def get_type_annotation(self, _: click.Parameter) -> type[str | ExplicitNullType]:
    return type[str | ExplicitNullType]
```

But I'm not trying to get stuck on this, just understand what's possible / specific.

Reply:
**Code**
```python
    def get_type_annotation(self, _: click.Parameter) -> type[str | ExplicitNullType]:
        return type[str | ExplicitNullType]
```

**Mypy Errors**
```
error: Returning Any from function declared to return "type[str] | type[ExplicitNullType]"  [no-any-return]
```

```
error: The type "type[type]" is not generic and not indexable  [misc]
```

Reply:
I would not want to change the runtime value which is returned here without a more deliberate process about getting that new style working in the click type checking. `click-type-test` is expecting the `types.UnionType` value to be returned, and isn't expecting an instance of `type.__class_getitem__` (does that produce a `typing.Type`?).

`click-type-test` would need to be updated with whatever change and potentially use `get_origin` or `get_args` to traverse the value being returned. It's something which could possibly get cleaned up, but I think that's a bigger conversation to be had.

And it may turn out that when we go to cases like this one to cleanup, we find that we can remove `get_type_annotation` hook entirely. The [v0.0.6](https://github.com/sirosen/click-type-test/releases/tag/0.0.6) features should cover most of our cases (which is why this hook is gone from most of our custom types).